### PR TITLE
Fix issue #457, and display hour 00 label if possible

### PR DESCRIPTION
**Pull request by rparkins999**
Fix issue #457  
The underlying cause was that mViewStartY could be set to a negative value when scrolling or attempting to scroll. It was getting forced to zero, but not until after a paint, hence the dithering. I had to abolish the overscroll bounce (which i don't consider to be a very nice or useful UI feature), since this is one of the cases where mViewStartY could be set to a negative value. It may be possible to fix issue #457 without abolishing the overscroll bounce, but this is beyond my ability to understand the code.

This pull request also contains a commit which arranges to display the label for hour 00 if there is room to do so, which there always is in the week view, or in the day view, if either there are any all day events or I can make use of the extra bit of space left over because the hour grid doesn't in general fill the available height.

### CODE REVIEW

**Comment by jspricke** [comment]
Thanks for looking into this!
You PR contains a number of unrelated commits (merges..) Can you clean it up?
If you are not familiar with git, you can give me access to your branch and I can do it for you.

**Comment by rparkins999** [comment]
Most of the other commits are to bring my original fork up to date with  Etar-Group /Etar-Calendar . The only unrelated commit is to display hour 00 when possible, which I don't think is objectionable.

**Comment by jspricke** [comment]
But they are all part of your pull request making it hard to review. Look here: https://github.com/Etar-Group/Etar-Calendar/pull/692/files

File: src/com/android/calendar/DayView.java [hunk 1]
```diff
@@ -785,6 +787,7 @@ private static int getEventAccessLevel(Context context, Event e) {
         return ACCESS_LEVEL_DELETE;
     }
 
+    @SuppressLint("MissingSuperCall")
```
Comment (jspricke) [inline src/com/android/calendar/DayView.java:790]: why don't you add `super` instead?

File: src/com/android/calendar/DayView.java [hunk 2]
```diff
@@ -1084,10 +1087,16 @@ public void setSelected(Time time, boolean ignoreTime, boolean animateToday) {
...
-
-        mViewStartY = viewStartY;
+        else if (viewStartY < 0) {
+            mViewStartY = 0;
+        }
+        else
+        {
+            mViewStartY = viewStartY;
+        }
+        if (DEBUG_SCROLLING) Log.e(TAG, "setViewStartY 1099: mViewStartY=" + mViewStartY);
```
Comment (jspricke) [inline src/com/android/calendar/DayView.java:1099]: Does it make sense to leave the debug code in there?

File: src/com/android/calendar/DayView.java [hunk 3]
```diff
@@ -2129,6 +2152,7 @@ private void computeEventRelations() {
         initAllDayHeights();
     }
 
+    @SuppressLint("WrongCall")
```
Comment (jspricke) [inline src/com/android/calendar/DayView.java:2155]: that sounds wrong, why do you do it?

**Comment by rparkins999** [comment]
On Thursday, 7 May 2020, 22:15:00 BST, Jochen Sprickerhof <notifications@github.com> wrote:  
 
 


@jspricke commented on this pull request.

In src/com/android/calendar/DayView.java:
> @@ -785,6 +787,7 @@ private static int getEventAccessLevel(Context context, Event e) {
         return ACCESS_LEVEL_DELETE;
     }
 
+    @SuppressLint("MissingSuperCall")
why don't you add super instead?>> because it wasn't originally there, I presume for a good reason.>> my version of Android Studio presumably has this inspection enabled and>> whoever was building it before didn't, so I needed the SuppressLint to build it


In src/com/android/calendar/DayView.java:
>          }
-
-        mViewStartY = viewStartY;
+        else if (viewStartY < 0) {
+            mViewStartY = 0;
+        }
+        else
+        {
+            mViewStartY = viewStartY;
+        }
+        if (DEBUG_SCROLLING) Log.e(TAG, "setViewStartY 1099: mViewStartY=" + mViewStartY);
Does it make sense to leave the debug code in there?>>There's plenty of other debug code which does conditional logging in the same way.
>> DEBUG_SCROLLING is defined as false, so in a release build the test will be optimised away.>> I left it there because I might need it again if someone finds a bug in my changes.

In src/com/android/calendar/DayView.java:
> @@ -2129,6 +2152,7 @@ private void computeEventRelations() {
         initAllDayHeights();
     }
 
+    @SuppressLint("WrongCall")
that sounds wrong, why do you do it>> Again I didn't change the code here.>> I needed the SuppressLint to make it compile. Android Studio doesn't like calls to onDraw().

**Comment by jspricke** [comment]
* rparkins999 <notifications@github.com> [2020-05-07 14:53]:
>In src/com/android/calendar/DayView.java:
>> @@ -785,6 +787,7 @@ private static int getEventAccessLevel(Context context, Event e) {
>         return ACCESS_LEVEL_DELETE;
>     }
>
>+    @SuppressLint("MissingSuperCall")
>why don't you add super instead?>> because it wasn't originally there, I presume for a good reason.>> my version of Android Studio presumably has this inspection enabled and>> whoever was building it before didn't, so I needed the SuppressLint to build it

Did you try adding the super call?

**Comment by rparkins999** [comment]
>> No, I didn't try adding the super call, because I don't have a full test suite and a full expected behaviour description which would enable me to discover if it broke anything.
>> According to git the override for onAttachedToWindow() was added by RoboErik <epastern@google.com> in commit f26ea9324d6d807b35ce68d89134ebb66c88706c on 17 November 2011, so I think he would be the right person to ask why it doesn't call through to the superclass implementation as the documentation for View.onAttachedToWindow() says it should.
>> If he is (or was then) a Google employee, he presumably knew what he was doing.

    On Friday, 8 May 2020, 10:41:42 BST, Jochen Sprickerhof <notifications@github.com> wrote:  
 
 

* rparkins999 <notifications@github.com> [2020-05-07 14:53]:
>In src/com/android/calendar/DayView.java:
>> @@ -785,6 +787,7 @@ private static int getEventAccessLevel(Context context, Event e) {
> return ACCESS_LEVEL_DELETE;
> }
>
>+ @SuppressLint("MissingSuperCall")
>why don't you add super instead?>> because it wasn't originally there, I presume for a good reason.>> my version of Android Studio presumably has this inspection enabled and>> whoever was building it before didn't, so I needed the SuppressLint to build it

Did you try adding the super call?

**Comment by rparkins999** [comment]
> I can't select it at all. It always jumps to hour 20. I've also tried it with the emulator. When I clicked with the mouse into the last field it always jumped to 8pm. For me it seems the last field isn't clickable at all.

I think this is issue #693. For me, sometimes it selects hour 20, and sometimes it selects hour 23. It seems to depend on the finger lift. As you lift your finger off the screen, the centre of the touched region can move: the software can interpret this as a scrolling action.

**Comment by SebiderSushi** [comment]
### Thank you very much for taking your time to fix this bug!

Now, all of this looks pretty rich and there are a lot of changes that do not need to be part of this pull request.

It looks a lot as if only the changes to `src/com/android/calendar/DayView.java` are relevant for the pull request.
If i haven't missed anything, the commits doing the needed changes are 36f1e6609eea42a4a8c5ea942983e577dea4a484, cd1ecb86bba786d5d7905319298dad1af0990375, b043785f79d46e6f24035ad0b6a51c5216bae41d, 8f59dc15ed7fc97edcac7d1167696e9aa8dc27b0 and dbfed92215570932c7acb4d2d2b4aa7b4e255e3a.
 
excerpt from `git log --pretty=oneline rparkins999/master`:
```
dbfed92215570932c7acb4d2d2b4aa7b4e255e3a Fix problem with selecting hour 23
8f59dc15ed7fc97edcac7d1167696e9aa8dc27b0 Don't mess with selected hour when scrolling
b043785f79d46e6f24035ad0b6a51c5216bae41d Remove DEBUG_SCROLLING, left in branch DEBUG_SCROLLING
cd1ecb86bba786d5d7905319298dad1af0990375 Display hour 00 label if there is room for it.
36f1e6609eea42a4a8c5ea942983e577dea4a484 Fix issue #457 and remove other bounces
```

`Fix problem with selecting hour 23` seems to deal with a problem that does not exist before applying `Fix issue #457 and remove other bounces`, so it could be squashed into the according commit.
`Remove DEBUG_SCROLLING, left in branch DEBUG_SCROLLING` can also be squashed into commit `Fix issue #457 and remove other bounces`.

`Fix issue #457 and remove other bounces` also contains some changes that i'd consider out of scope in this case, namely:
https://github.com/Etar-Group/Etar-Calendar/pull/692/files?file-filters%5B%5D=.java#diff-768137d9bb90169c844dd840cce00007R24
https://github.com/Etar-Group/Etar-Calendar/pull/692/files?file-filters%5B%5D=.java#diff-768137d9bb90169c844dd840cce00007R788
https://github.com/Etar-Group/Etar-Calendar/pull/692/files?file-filters%5B%5D=.java#diff-768137d9bb90169c844dd840cce00007R1119
https://github.com/Etar-Group/Etar-Calendar/pull/692/files?file-filters%5B%5D=.java#diff-768137d9bb90169c844dd840cce00007R2147
https://github.com/Etar-Group/Etar-Calendar/pull/692/files?file-filters%5B%5D=.java#diff-768137d9bb90169c844dd840cce00007R4940
That is, at `DayView.java`, lines `24`, `788`, `1119`, `2147` and `4940`, after applying commit `Fix issue #457 and remove other bounces`.

I'm not sure whether to think of `Display hour 00 label if there is room for it.` as a standalone change. I can imagine both this or squashing it into `Fix issue #457 and remove other bounces` as well.

To me, `Don't mess with selected hour when scrolling` looks like a very appealing feature. Other than that, i am unsure whether it should be included in this specific pull request.

### The following is what i personally would do to produce a clean history in this case:
(not all steps might be necessary)
```
# origin is the name to the remote of your own github fork
cd Etar-Calendar
git remote add upstream https://github.com/Etar-Group/Etar-Calendar
git fetch upstream
git fetch origin
git checkout -b fix_dithering_dayview upstream/master
git cherry-pick 36f1e66 b043785 dbfed92 cd1ecb8 8f59dc1
git rebase -i upstream/master
# Read the hints on screen,
# edit 36f1e66
# squash or fixup b043785, dbfed92 and cd1ecb8
#
# after the rebase has interrupted to edit commit 36f1e66, undo the out of scope changes, then
git commit --amend --no-edit src/com/android/calendar/DayView.java
git rebase --continue
git push --force origin fix_dithering_dayview:master
```
The force push forces the new history onto your GitHub master branch, as this is the one linked to this Pull Request right now, and i don't think this can be changed without creating a new PR. As soon as the Request is merged into this upstream repo, you can delete or force push anything you want to your own GitHub master branch again.

You can also skip the cherry pick and paste the commit hashes into the editor when rebasing. The following rebase todo list also does the cherry picking part:
```
e 36f1e66
f b043785
f dbfed92
f cd1ecb8
p 8f59dc1
```


### Summary
As you can see, the PR ended up with a diff of only 2 to 3 commits changing one single file and implementing 2 (or 3) fixes and features, depending on how you handle `Display hour 00 label if there is room for it.`.
This is how it looks: https://github.com/SebiderSushi/Etar-Calendar

**Comment by SebiderSushi** [comment]
I did some testing and apparently, `dbfed92215570932c7acb4d2d2b4aa7b4e255e3a Fix problem with selecting hour 23` only works if applied together with `8f59dc15ed7fc97edcac7d1167696e9aa8dc27b0 Don't mess with selected hour when scrolling`.

Now i am very confused as to how to squash of retain the commits of this PR.

**Comment by jspricke** [comment]
@SebiderSushi I had a quick look into your branches and https://github.com/SebiderSushi/Etar-Calendar/commits/squashed looks ok to go forward (I didn't do a complete review). Can you send a PR with it?

**Comment by SebiderSushi** [comment]
If you insist, i will squash my master branch however you like since you're a project member and thus one of the people who have to deal with it.
However, i don't think the `squashed` branch is ideal so i will quickliy open a PR from what is my current master branch, because it will enable code review while still being the most verbose and true to history, i.e. no commits have been melded yet, it's just the cherry-picks. My suggestion is to go from there. Whatever will be decided for the final squash that should be merged, i would definitely change the commit messages a bit to reflect on the squashed changes which, as i know now, are quite intertwined.

**Comment by SebiderSushi** [comment]
Note that the net-diff is the same across all of my branches. They are just differently squashed versions of the master branch.

**Comment by jspricke** [comment]
Merged via #710, thanks for your work on this!

**Pull request closed without merging**